### PR TITLE
Validate storage medium for empty dir

### PR DESCRIPTION
**Pull request by xiangpengzhao**
**What this PR does / why we need it**:
We should validate the medium type of empty dir.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/api/types.go [hunk 1]
```diff
@@ -686,6 +687,12 @@ const (
 	StorageMediumHugePages StorageMedium = "HugePages" // use hugepages
 )
 
+var ValidStorageMedia = sets.NewString(
+	string(StorageMediumDefault),
+	string(StorageMediumMemory),
+	string(StorageMediumHugePages),
+)
+
```
Comment (krmayankk) [inline pkg/api/types.go]: this doesnt look like it should go in types.go

Comment (xiangpengzhao) [reply pkg/api/types.go]: Good point! Addressed. PTAL. Thanks!

**Comment by derekwaynecarr** [comment]
good catch.

/lgtm
/approve no-issue

**Comment by xiangpengzhao** [comment]
/assign @dchen1107
for final approval, thanks!

**Comment by xiangpengzhao** [comment]
kindly ping api-approvers @liggitt @erictune

**Comment by liggitt** [comment]
what's the current behavior if an invalid medium is set?

if a pod currently has an invalid medium, and this validation tightens, does it block deletion of the bad pod?

**Comment by xiangpengzhao** [comment]
>what's the current behavior if an invalid medium is set?

@liggitt I remember that months ago my user encountered a weird issue during the container starting in such case. I will try to reproduce it tomorrow.

>if a pod currently has an invalid medium, and this validation tightens, does it block deletion of the bad pod?

As is mentioned in my above sentence, the pod won't be started successfully, so the bad pod won't be blocked when being deleted, IMO.

**Comment by liggitt** [comment]
Deletion can require an API update in order to remove a finalizer. Will this validation change block that update? If so, it would prevent deletion.

**Comment by xiangpengzhao** [comment]
>Deletion can require an API update in order to remove a finalizer. Will this validation change block that update? If so, it would prevent deletion.

Not sure about this yet. I will dig into the code or reproduce it :)

**Comment by xiangpengzhao** [comment]
>if a pod currently has an invalid medium, and this validation tightens, does it block deletion of the bad pod?

@liggitt how to produce this case? Should I create a pod with an invalid medium against a kube-apiserver without the validation, and then restart kube-apiserver by the one with this validation, and try to delete the bad pod? (can this happen?)

**Comment by liggitt** [comment]
> @liggitt how to produce this case? Should I create a pod with an invalid medium against a kube-apiserver without the validation

Yes, with a finalizer

> and then restart kube-apiserver by the one with this validation, and try to delete the bad pod? (can this happen?)

Yes. Before the pod can be deleted, the finalizer must be removed, which requires an update, which requires the pod spec pass validation.

**Comment by xiangpengzhao** [comment]
@liggitt the deletion is prevented by the validation as you were thinking of. So the validation is a break change, right? Then should we tighten the validation? I guess we still should, and maybe we can write an `action required` release note for this.

Deletion is prevented:
```
# ./kubectl delete rc rc-nginx-single
error: Scaling the resource failed with: ReplicationController "rc-nginx-single" is invalid: [spec.template.spec.volumes[0].emptyDir.medium: Invalid value: "invalidmed": invalid storage medium for EmptyDir volumes, spec.template.spec.containers[0].volumeMounts[0].name: Not found: "test-volume"]; Current resource version 30667
```

BTW, w/o this validation, the container mounting the invalid volume cannot be started, and the pod will be pending in `ContainerCreating` state.
```
Events:
  Type     Reason       Age               From                      Message
  ----     ------       ----              ----                      -------
  Normal   Scheduled    18s               default-scheduler         Successfully assigned rc-nginx-single-4ptbf to 192.168.122.152
  Warning  FailedMount  2s (x6 over 18s)  kubelet, 192.168.122.152  MountVolume.SetUp failed for volume "test-volume" : unknown storage medium "invalidmed"
```

**Comment by liggitt** [comment]
What action required would you suggest? Some upgrade procedures require cycling active objects to ensure storage migration. This would break those upgrades, in order to make one type of failing pod fail in a different way. That doesn't seem like a good trade off.

**Comment by xiangpengzhao** [comment]
I'd suggest something like `If your cluster has pods with invalid medium type emptydir, be sure to delete those pods before upgrading and then correct the medium type and recreate the pods`.

Since the pods with invalid medium type aren't running and serving before upgrading, I think it's safe to delete them. OTOH, making pods with invalid medium type is user error. This is not expected and should be corrected. Also, IMO, it's better to validate and error out earlier than later :)

**Comment by liggitt** [comment]
> Since the pods with invalid medium type aren't running and serving before upgrading, I think it's safe to delete them.

Safe to delete, sure. Safe to make undeletable if a release note was missed, no.

> OTOH, making pods with invalid medium type is user error. This is not expected and should be corrected. Also, IMO, it's better to validate and error out earlier than later :)

Preventing such pods from being created seems ok. Preventing them from being deleted does not.

**Comment by xiangpengzhao** [comment]
@liggitt So I guess you are inclined to leave the code here as-is (forever :smile: ) as per the concern. It would be ok if so, since kubelet will detect this (`MountVolume.SetUp failed for volume "test-volume" : unknown storage medium "invalidmed"`) and report to users.

**Comment by liggitt** [comment]
> @liggitt So I guess you are inclined to leave the code here as-is (forever 😄 ) as per the concern

or make the validation change only apply on create, not update

**Comment by xiangpengzhao** [comment]
>or make the validation change only apply on create, not update

SGTM :+1:  This will need to change signatures (i.e., add a boolean param `isUpdate` or something) of some associated Validate funcs, SGTY? @liggitt

**Comment by liggitt** [comment]
> This will need to change signatures (i.e., add a boolean param isUpdate or something) of some associated Validate funcs

possibly... the ripples could get large quickly

**Comment by liggitt** [comment]
/hold

c.f. https://github.com/kubernetes/kubernetes/issues/59984

**Comment by derekwaynecarr** [comment]
inclined to change nothing right now given all issues discussed.

**Comment by xiangpengzhao** [comment]
Agreed.

**Pull request closed without merging**